components/bpmn/org.wso2.carbon.bpmn.analytics.publisher/src/main/java/org/wso2/carbon/bpmn/analytics/publisher/utils/AnalyticsPublishServiceUtils.java

```diff
@@ -0,0 +1,301 @@
+			bpmnProcessInstance.setTenantId(instance.getTenantId());
+			bpmnProcessInstance.setName(instance.getName());
+			bpmnProcessInstance.setInstanceId(instance.getId());
+			bpmnProcessInstance.setBusinessKey(instance.getBusinessKey());
+			bpmnProcessInstance.setStartTime(instance.getStartTime());
+			bpmnProcessInstance.setEndTime(instance.getEndTime());
+			bpmnProcessInstance.setDuration(instance.getDurationInMillis());
+			bpmnProcessInstance.setStartUserId(instance.getStartUserId());
+			bpmnProcessInstance.setStartActivityId(instance.getStartActivityId());
+			bpmnProcessInstance.setVariables(formatVariables(instance.getProcessVariables()));
+			bpmnProcessInstances.add(bpmnProcessInstance);
+		}
+		return bpmnProcessInstances.toArray(new BPMNProcessInstance[bpmnProcessInstances.size()]);
+	}
+
+	/**
+	 * convert historic task instances to BPMN task instances
+	 *
+	 * @param historicTaskInstanceList List of historic task instances
+	 * @return BPMNTaskInstance array
+	 */
+	private BPMNTaskInstance[] getBPMNTaskInstances(
+			List<HistoricTaskInstance> historicTaskInstanceList) {
+		BPMNTaskInstance bpmnTaskInstance;
+		List<BPMNTaskInstance> bpmnTaskInstances = new ArrayList<>();
+		for (HistoricTaskInstance taskInstance : historicTaskInstanceList) {
+			bpmnTaskInstance = new BPMNTaskInstance();
+			bpmnTaskInstance.setTaskDefinitionKey(taskInstance.getTaskDefinitionKey());
+			bpmnTaskInstance.setTaskInstanceId(taskInstance.getId());
+			bpmnTaskInstance.setAssignee(taskInstance.getAssignee());
+			//claim time is not stored in the DB
+			//bpmnTaskInstance.setClaimTime(taskInstance.getClaimTime());
+			bpmnTaskInstance.setStartTime(taskInstance.getStartTime());
+			bpmnTaskInstance.setEndTime(taskInstance.getEndTime());
+			bpmnTaskInstance.setTaskName(taskInstance.getName());
+			bpmnTaskInstance.setDurationInMills(taskInstance.getDurationInMillis());
+			//bpmnTaskInstance.setWorkTimeInMills(taskInstance.getWorkTimeInMillis());
+			bpmnTaskInstance.setCreateTime(taskInstance.getCreateTime());
+			bpmnTaskInstance.setOwner(taskInstance.getOwner());
+			bpmnTaskInstance.setProcessInstanceId(taskInstance.getProcessInstanceId());
+			bpmnTaskInstances.add(bpmnTaskInstance);
+		}
+		return bpmnTaskInstances.toArray(new BPMNTaskInstance[bpmnTaskInstances.size()]);
+	}
+
+	/**
+	 * Format the process instance variables as BPMNVariable objects
+	 *
+	 * @param processVariables variables which belongs to a given process instance as key, value pairs
+	 * @return BPMNVariable objects array
+	 */
+	private BPMNVariable[] formatVariables(Map<String, Object> processVariables) {
+		if (processVariables == null) {
+			return null;
+		}
+		BPMNVariable[] vars = new BPMNVariable[processVariables.size()];
+		int currentVar = 0;
+		for (Map.Entry entry : processVariables.entrySet()) {
+			vars[currentVar] = new BPMNVariable(entry.getKey().toString(),
+			                                    processVariables.get(entry.getKey().toString())
```

Comment:
Here Map.Entry is used to get the each entry in the processVariables.entrySet() and from that fill the  BPMNVariable[] aray because we need to give an array of variable data to the publisher.